### PR TITLE
FIXES #1268: provides array based paginator

### LINKED ISSUE #1268: A Paginator Adapter class that works from an array
Current paginators work from a sql query.
Required to resolve issue #1256.
Please assign to me.

### PR DESCRIPTION
**Pull request by scottcwilson**

### CODE REVIEW

File: includes/library/zencart/Paginator/src/adapters/ArrayFactory.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        return $resultList;
+    }
+
+    /**
+     * @param $data
+     * @return mixed
+     */
+    public function getTotalItemCount($data)
+    {
+        $totalsize = count($data);
```
Comment (zcwilt) [inline includes/library/zencart/Paginator/src/adapters/ArrayFactory.php]: needs to return the value

File: includes/library/zencart/Paginator/src/adapters/ArrayFactory.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+namespace ZenCart\Paginator\adapters;
+
+use ZenCart\Paginator\AdapterInterface;
+use ZenCart\Paginator\AbstractAdapter;
+
+/**
+ * Class ArrayFactory
+ * @package ZenCart\Paginator\adapters
+ */
+class ArrayFactory extends AbstractAdapter implements AdapterInterface
```
Comment (zcwilt) [inline includes/library/zencart/Paginator/src/adapters/ArrayFactory.php]: Note the class isn't a Factory, it's my fault for calling the other adapter QueryFactory.
I called it that as it is operating on a QueryFactory($db) sql query.

Comment (zcwilt) [inline includes/library/zencart/Paginator/src/adapters/ArrayFactory.php]: Maybe better to call the class ArrayData. Will rename the QueryFactory adapter to remove ambiguity.

Comment (scottcwilson) [inline includes/library/zencart/Paginator/src/adapters/ArrayFactory.php]: Should we merge this and then I'll do another update for the renames?

Comment (zcwilt) [inline includes/library/zencart/Paginator/src/adapters/ArrayFactory.php]: I have done https://github.com/zencart/zencart/pull/1280
as it also needed some unit test changes.

Comment (zcwilt) [inline includes/library/zencart/Paginator/src/adapters/ArrayFactory.php]: so would just do your rename in this pr

**Comment by zcwilt** [comment]
:+1:

File: includes/library/zencart/Paginator/src/adapters/ArrayData.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ */
+class ArrayData extends AbstractAdapter implements AdapterInterface
+{
+
+    /**
+     * @param $data
+     * @param array $params
+     * @return array
+     */
+    public function getResultList($data, array $params)
```
Comment (drbyte) [inline includes/library/zencart/Paginator/src/adapters/ArrayData.php]: I realize the `AdapterInterface` signature doesn't specify `$data` to be an array, but if this adapter is requiring arrays, should it be typehinted in the signature, the same as `$params` is?

File: includes/library/zencart/Paginator/src/adapters/ArrayData.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        $len = $params['itemsPerPage']; 
+        $resultList = array_slice($data, $start, $len); 
+        return $resultList;
+    }
+
+    /**
+     * @param $data
+     * @return mixed
+     */
+    public function getTotalItemCount($data)
```
Comment (drbyte) [inline includes/library/zencart/Paginator/src/adapters/ArrayData.php]: In this case, given that we run `count($data)`, should we typehint `$data` to be `Countable`?

**Comment by drbyte** [comment]
Hmmm ... Those suggestions would require a change to the whole Interface, and to other adapters.
Will leave it for now.

👍

**Pull request merged**